### ISSUE TITLE
app.modeller.addSolvent to rotate protein and minimize volume 

### ISSUE THREAD

**Issue by rafwiewiora** [pandegroup/openmm#1667]
While setting up a number of simulations recently of the same protein sequence in slightly different conformations, which had not been superposed, I noticed that given a choice of padding in `app.modeller.addSolvent()` one can get large variations in the number of waters / volume that results. 

For example in the FAH projects here: https://github.com/choderalab/fah-projects/issues/55 I went from 52k atoms to 34k atoms by superposing the 'problematic' structure onto another which was in a less-waters-needed rotation.

It would be great if `addSolvent` rotated the protein to minimize the volume of the box, given a choice of padding. What do you think? 

Tagging @jchodera

**Comment by peastman** [comment]
I can think of a few ways of trying to do that, though none of them is exact.  You could just compute the principle axes of the inertia tensor and align them with the coordinate axes.  That will be an approximation, but hopefully not too bad, and it would be very fast.  Or you could use an optimizer to explicitly search for a minimum of the box volume, although that could easily get stuck in a local minimum.  Maybe first do one, then the other?

One thing to be aware of is that you don't necessarily really want this.  The goal of padding is to make sure no part of the protein ever comes too close to a periodic copy of itself.  If the protein can freely rotate, you actually want to find the orientation that *maximizes* the box volume rather than minimizing it.  Otherwise, there's a danger that the protein could rotate and collide with itself.  That's mainly a problem if the protein is much wider in some directions than in others.

**Comment by zhang-ivy** [comment]
Recently, @glass-w and I have been running simulations on Folding@home on covid systems that are very long in one direction. We were using a rectangular box and realized that after awhile, the complex rotates such that the long axis of the complex is oriented along the short axis of the rectangular box. This caused the system to interact with the image of itself, so we ended up switching to a truncated octahedron. We were parametrizing using amber tleap (because OpenMM doesn't currently support glycan forcefields yet), so switching the box to a truncated octahedron was easy. Just want to reiterate that allowing for non-cubic boxes would be really helpful!

**Comment by peastman** [comment]
Coincidentally, someone was just asking on the forum about how to prevent a protein from rotating.  It might be relevant to you.

https://simtk.org/plugins/phpBB/viewtopicPhpbb.php?f=161&t=13026&p=0&start=0&view=&sid=2d72af7885dadcf906c58c78d9882fa3

> Just want to reiterate that allowing for non-cubic boxes would be really helpful!

Allowing for them in what context?  `addSolvent()` already supports non-cubic boxes by using either the `boxSize` or `boxVectors` argument.  OpenMM-Setup also allows you to enter any box dimensions you want, or it will automatically build a truncated octahedron if that's what you want.

**Comment by zhang-ivy** [comment]
>it will automatically build a truncated octahedron if that's what you want.
Can I specify that I want it to build a truncated octahedron with X padding without specifying the box vectors?

**Comment by peastman** [comment]
Yes.  When you choose to specify a padding distance, there's a menu for what box shape you want: cube, truncated octahedron, or rhombic dodecahedron.

**Comment by glass-w** [comment]
>you actually want to find the orientation that maximizes the box volume rather than minimizing it. Otherwise, there's a danger that the protein could rotate and collide with itself. That's mainly a problem if the protein is much wider in some directions than in others.

Maybe it would be useful to have a warning to indicate if your system may undergo self interaction? At the moment it seems there isn't such a warning. I'm not sure how feasible this is, but it would be a good addition.

**Comment by jchodera** [comment]
+1 for warnings!

**Comment by peastman** [comment]
In the current implementation, there's nothing to warn about.  It picks the box size based on the largest dimension of your molecule, so it won't interact with itself however it rotates.  If we implemented this feature, there would almost *always* be a risk of self interactions, since the whole point is to make the box smaller in some dimensions than others.
